core-services/src/test/java/org/zalando/nakadi/validation/ValidationHelperServiceTest.java

```diff
@@ -0,0 +1,157 @@
+            "test/" + LARGE_NAME + "z", // Key name cannot be more than 63 characters
+            "A-test/test", // Prefix cannot have upper case
+            ".test/test", // Prefix should be valid subdomain
+            "test..com/test", // Prefix should be valid subdomain
+            "test.-org/test", // Prefix should be valid subdomain
+            LONG_PREFIX + "z/test" // Key prefix cannot be more than 253 characters
+    );
+
+    private static final List<String> VALID_KEYS = Arrays.asList(
+            "test-annotation",
+            "nakadi.io/test-annotation",
+            "nakadi-test.io/test-annotation",
+            "nakadi-test/test-Annotation",
+            "test/t-e_s.t",
+            "test/1t-e_s.t0",
+            LONG_PREFIX + "/" + LARGE_NAME
+    );
+
+    private final ResourceValidationHelperService validationHelperService = new ResourceValidationHelperService();
+
+    @Test
+    public void testCheckAnnotationPassForValidAnnotations() {
+        validationHelperService.checkAnnotations(null);
+        final ResourceAnnotations annotations = new ResourceAnnotations();
+        for (final String validKey : VALID_KEYS) {
+            annotations.put(validKey, "testValue");
+        }
+        try {
+            validationHelperService.checkAnnotations(annotations);
+        } catch (final InvalidResourceAnnotationException irae) {
+            Assert.fail("Validation failed with message" + irae.getMessage());
+        }
+    }
+
+    @Test
+    public void testCheckAnnotationFailsForInvalidAnnotationKeys() {
+        for (final String invalidKey : INVALID_KEYS) {
+            final ResourceAnnotations annotations = new ResourceAnnotations();
+            annotations.put(invalidKey, "invalidAnnotation");
+            Assert.assertThrows(
+                    "Invalid key " + invalidKey + " passed validation",
+                    InvalidResourceAnnotationException.class,
+                    () -> validationHelperService.checkAnnotations(annotations));
+        }
+    }
+
+    @Test
+    public void testCheckAnnotationFailsForLongerAnnotationValue() {
+        final ResourceAnnotations annotations = new ResourceAnnotations();
+        final String longerAnnotation = "a".repeat(ResourceValidationHelperService.MAX_ANNOTATION_LENGTH + 1);
+        annotations.put("nakadi.io/test", longerAnnotation);
+        Assert.assertThrows(
+                "Invalid key annotation value passed validation",
+                InvalidResourceAnnotationException.class,
+                () -> validationHelperService.checkAnnotations(annotations));
+    }
+
+    @Test
+    public void testCheckAnnotationPassForValidAnnotationValues() {
+        final ResourceAnnotations annotations = new ResourceAnnotations();
+        final String longerAnnotation = "a".repeat(ResourceValidationHelperService.MAX_ANNOTATION_LENGTH);
+        annotations.put("nakadi.io/test-long", longerAnnotation);
+        annotations.put("nakadi.io/test-null", null);
```

Comment:
Wonder how nulls are re-serialized?  Does it become an empty string?

Reply:
It will be `"nakadi.io/test-null" : null` in this case.